File: test/PairingTest.kt
#!/usr/bin/env kotlin

/**
 * Standalone Pairing Test Script
 * Run with: kotlin PairingTest.kt
 * Requires: Kotlin compiler + BouncyCastle
 *
 * Or compile and run:
 *   kotlinc -include-runtime -d pairing_test.jar PairingTest.kt
 *   java -jar pairing_test.jar
 */

import java.io.*
import java.net.*
import java.security.*
import java.security.spec.*
import java.util.*
import javax.crypto.*
import javax.crypto.spec.*
import org.bouncycastle.jce.provider.BouncyCastleProvider()
import org.bouncycastle.util.encoders.Base64

// Install BouncyCastle
Security.addProvider(BouncyCastleProvider())

println("=== IINA Remote Pairing Test ===")
println()

// Configuration - EDIT THESE VALUES
var HOST = "192.168.1.100"
var PORT = 8765
var PAIR_CODE = "123456"
var DEVICE_NAME = "Test Device"

fun log(msg: String) {
    println("[LOG] $msg")
}

fun error(msg: String) {
    println("[ERROR] $msg")
}

// Parse command line args
args.forEachIndexed { index, arg ->
    when (arg) {
        "-h" -> HOST = args.getOrElse(index + 1) { HOST }
        "-p" -> PORT = args.getOrElse(index + 1) { "8765" }.toIntOrNull() ?: 8765
        "-c" -> PAIR_CODE = args.getOrElse(index + 1) { "123456" }
        "-n" -> DEVICE_NAME = args.getOrElse(index + 1) { DEVICE_NAME }
        "--help" -> {
            println("Usage: kotlin PairingTest.kt [options]")
            println("Options:")
            println("  -h <host>    Server host (default: $HOST)")
            println("  -p <port>    Server port (default: $PORT)")
            println("  -c <code>    Pair code (default: $PAIR_CODE)")
            println("  -n <name>    Device name (default: $DEVICE_NAME)")
            println()
            println("Example:")
            println("  kotlin PairingTest.kt -h 192.168.1.100 -p 8765 -c ABC123")
            System.exit(0)
        }
    }
}

log("Configuration:")
log("  Host: $HOST:$PORT")
log("  Code: $PAIR_CODE")
log("  Name: $DEVICE_NAME")
println()

// Generate Ed25519 key pair
log("Step 1: Generating Ed25519 key pair...")
val keyGen = KeyPairGenerator.getInstance("Ed25519")
val keyPair = keyGen.generateKeyPair()
val publicKey = keyPair.public
val privateKey = keyPair.private
val publicKeyBytes = publicKey.encoded
val publicKeyB64 = Base64.toBase64String(publicKeyBytes)
log("  Public key (base64): ${publicKeyB64.take(32)}...")
log("  Key generated successfully!")
println()

// Generate device ID
val deviceId = UUID.randomUUID().toString()
log("Step 2: Generated Device ID: $deviceId")
println()

// HTTP request helper
fun httpGet(path: String): String? {
    return try {
        val url = URL("http://$HOST:$PORT$path")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.connectTimeout = 10000
        conn.readTimeout = 10000
        conn.connect()
        val response = conn.inputStream.bufferedReader().readText()
        log("  GET $path -> ${conn.responseCode}")
        conn.disconnect()
        response
    } catch (e: Exception) {
        error("  GET $path failed: ${e.message}")
        null
    }
}

fun httpPost(path: String, body: String): String? {
    return try {
        val url = URL("http://$HOST:$PORT$path")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json")
        conn.connectTimeout = 10000
        conn.readTimeout = 10000
        conn.outputStream.write(body.toByteArray())
        conn.outputStream.flush()
        val response = if (conn.responseCode < 400) {
            conn.inputStream.bufferedReader().readText()
        } else {
            conn.errorStream?.bufferedReader()?.readText() ?: ""
        }
        log("  POST $path -> ${conn.responseCode}")
        if (conn.responseCode >= 400) {
            log("  Error response: $response")
        }
        conn.disconnect()
        response
    } catch (e: Exception) {
        error("  POST $path failed: ${e.message}")
        null
    }
}

// Get pair challenge
log("Step 3: Getting pair challenge...")
val challengeJson = httpGet("/api/v1/pair/challenge")
if (challengeJson == null) {
    error("Failed to get challenge!")
    System.exit(1)
}
log("  Challenge response: $challengeJson")
println()

// Submit pair code
log("Step 4: Submitting pair code...")
val deviceIdJson = "\"$deviceId\""
val deviceNameJson = "\"$DEVICE_NAME\""
val publicKeyJson = "\"$publicKeyB64\""
val pairRequestJson = """
{
    "code": "$PAIR_CODE",
    "device_id": "$deviceId",
    "device_name": "$DEVICE_NAME",
    "public_key": "$publicKeyB64"
}
""".trimIndent()
log("  Request: $pairRequestJson")
val pairResponseJson = httpPost("/api/v1/pair", pairRequestJson)
if (pairResponseJson == null) {
    error("Failed to submit pair code!")
    System.exit(1)
}
log("  Pair response: $pairResponseJson")
println()

// Parse response
try {
    // Simple JSON parsing without Gson
    val tokenMatch = Regex("""\"token"\s*:\s*"([^"]+)"""").find(pairResponseJson)
    val serverNameMatch = Regex("""\"server_name"\s*:\s*"([^"]+)"""").find(pairResponseJson)
    val expiresInMatch = Regex("""\"expires_in"\s*:\s*(\d+)""").find(pairResponseJson)

    val token = tokenMatch?.groupValues?.get(1)
    val serverName = serverNameMatch?.groupValues?.get(1)
    val expiresIn = expiresInMatch?.groupValues?.get(1)?.toIntOrNull()

    if (token != null) {
        println()
        println("=== PAIRING SUCCESSFUL ===")
        log("  Token: ${token.take(20)}...")
        log("  Server Name: $serverName")
        log("  Expires In: $expiresIn seconds")
    } else {
        println()
        println("=== PAIRING FAILED ===")
        log("  Response: $pairResponseJson")
    }
} catch (e: Exception) {
    error("Failed to parse response: ${e.message}")
}

File: test/PairingTest.py
#!/usr/bin/env python3
"""
IINA Remote Pairing Test
Tests the complete pairing flow with Ed25519 keys.

Usage:
    python3 PairingTest.py -h <host> -p <port> -c <code> [-n <name>]
"""

import argparse
import base64
import json
import sys
import urllib.request
import urllib.error
import uuid

try:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    from cryptography.hazmat.primitives import serialization
    HAS_CRYPTO = True
except ImportError:
    HAS_CRYPTO = False
    print("[ERROR] cryptography library required: pip install cryptography")
    sys.exit(1)

def log(msg):
    print(f"[LOG] {msg}")

def error(msg):
    print(f"[ERROR] {msg}")

def http_get(host, port, path):
    url = f"http://{host}:{port}{path}"
    log(f"GET {url}")
    try:
        with urllib.request.urlopen(url, timeout=10) as resp:
            body = resp.read().decode('utf-8')
            log(f"  -> HTTP {resp.status}")
            return body
    except urllib.error.HTTPError as e:
        body = e.read().decode('utf-8') if e.fp else ""
        log(f"  -> HTTP {e.code} {e.reason}")
        if body:
            log(f"     Body: {body[:200]}")
        return None
    except Exception as e:
        error(f"  -> {e}")
        return None

def http_post(host, port, path, data):
    url = f"http://{host}:{port}{path}"
    log(f"POST {url}")
    try:
        req = urllib.request.Request(url,
            data=data.encode('utf-8'),
            headers={'Content-Type': 'application/json'},
            method='POST')
        with urllib.request.urlopen(req, timeout=10) as resp:
            body = resp.read().decode('utf-8')
            log(f"  -> HTTP {resp.status}")
            return body
    except urllib.error.HTTPError as e:
        body = e.read().decode('utf-8') if e.fp else ""
        log(f"  -> HTTP {e.code} {e.reason}")
        if body:
            log(f"     Body: {body[:500]}")
        return None
    except Exception as e:
        error(f"  -> {e}")
        return None

def main():
    parser = argparse.ArgumentParser(description='IINA Remote Pairing Test')
    parser.add_argument('--host', '-H', default='localhost', help='Server host')
    parser.add_argument('--port', '-P', type=int, default=8765, help='Server port')
    parser.add_argument('--code', '-c', required=True, help='Pair code (6 digits)')
    parser.add_argument('--name', '-n', default='TestDevice', help='Device name')
    args = parser.parse_args()

    print("=" * 60)
    print("IINA Remote Pairing Test")
    print("=" * 60)
    log(f"Host: {args.host}:{args.port}")
    log(f"Code: {args.code}")
    log(f"Name: {args.name}")
    print()

    # Step 1: Generate Ed25519 key pair (32-byte raw public key)
    log("Step 1: Generate Ed25519 key pair")
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()

    # Get raw 32-byte public key (this is what we send to server)
    public_bytes = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw
    )
    public_key_b64 = base64.b64encode(public_bytes).decode()
    log(f"  Public key (32 bytes, base64): {public_key_b64}")
    log(f"  Public key length: {len(public_bytes)} bytes")
    assert len(public_bytes) == 32, f"Public key must be 32 bytes, got {len(public_bytes)}"
    print()

    # Step 2: Generate device ID
    log("Step 2: Generate device ID")
    device_id = str(uuid.uuid4())
    log(f"  Device ID: {device_id}")
    print()

    # Step 3: Get pair challenge
    log("Step 3: Get pair challenge")
    challenge_resp = http_get(args.host, args.port, '/api/v1/pair/challenge')
    if challenge_resp is None:
        error("Failed to get challenge!")
        sys.exit(1)
    log(f"  Response: {challenge_resp}")
    print()

    # Step 4: Submit pair request with public key
    log("Step 4: Submit pair request")

    pair_request = {
        "code": args.code,
        "device_id": device_id,
        "device_name": args.name,
        "public_key": public_key_b64
    }
    request_body = json.dumps(pair_request)
    log(f"  Request body: {request_body}")

    pair_response = http_post(args.host, args.port, '/api/v1/pair', request_body)
    if pair_response is None:
        error("No response from server!")
        sys.exit(1)

    print()
    print("=" * 60)
    print("Response:")
    print(pair_response)
    print("=" * 60)

    # Parse response
    try:
        resp_json = json.loads(pair_response)
        if "token" in resp_json:
            print()
            print("=== PAIRING SUCCESSFUL ===")
            print(f"  Token: {resp_json['token'][:30]}...")
            print(f"  Server Name: {resp_json.get('server_name', 'N/A')}")
            print(f"  Expires In: {resp_json.get('expires_in', 'N/A')} seconds")
        else:
            print()
            print("=== PAIRING FAILED ===")
            print(f"  Response: {pair_response}")
    except json.JSONDecodeError:
        print()
        print("=== PAIRING FAILED (invalid JSON) ===")
        print(f"  Response: {pair_response}")

if __name__ == '__main__':
    main()
